### PR TITLE
[@types/pdfmake] Overhaul type definitions (cont. from #42787)

### PR DESCRIPTION
**Pull request by jeralm**
Continued from #42787 

This PR should address these issues: #42761  #35081 #31011
The goal of this PR was to bring the type definitions on par with every feature provided by pdfMake.
pdfMake is in large part documented through examples. I am happy to say that ALL examples type-check without errors under these new type definitions.

This might be a breaking change for some users, but the old definitions were not really usable and forced users to plague their code with `as any` type assertions.

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://pdfmake.github.io/docs/ http://pdfmake.org/playground.html https://github.com/bpampuch/pdfmake/tree/0.1/examples
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

File: types/pdfmake/interfaces.d.ts [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+    headerRows?: number;
+    dontBreakRows?: boolean;
+    keepWithHeaderRows?: number;
+    layout?: TableLayout;
+}
+
+export type PredefinedTableLayout = 'noBorders' | 'headerLineOnly' | 'lightHorizontalLines';
+export type TableLayout = PredefinedTableLayout | CustomTableLayout;
+
+export interface Style {
```
Comment (nmummau) [inline types/pdfmake/interfaces.d.ts]: Please add back type annotation comments.

File: types/pdfmake/index.d.ts [hunk 2]
```diff
@@ -5,289 +5,19 @@
...
-
-    type PageSize =
-        "4A0"|"2A0"|"A0"|"A1"|"A2"|"A3"|"A4"|"A5"|"A6"|"A7"|"A8"|"A9"|"A10"|
-        "B0"|"B1"|"B2"|"B3"|"B4"|"B5"|"B6"|"B7"|"B8"|"B9"|"B10"|
-        "C0"|"C1"|"C2"|"C3"|"C4"|"C5"|"C6"|"C7"|"C8"|"C9"|"C10"|
-        "RA1"|"RA2"|"RA3"|"RA4"|
-        "SRA1"|"SRA2"|"SRA3"|"SRA4"|
-        "EXECUTIVE"|"FOLIO"|"LEGAL"|"LETTER"|"TABLOID";
-
-    type PageOrientation = "portrait" | "landscape";
```
Comment (nmummau) [inline types/pdfmake/index.d.ts]: It looks like you did not add this type to the new interfaces.d.ts.

File: types/pdfmake/index.d.ts [hunk 3]
```diff
@@ -5,289 +5,19 @@
...
-
-    type Layout = "noBorders" | "headerLineOnly" | "lightHorizontalLines";
```
Comment (nmummau) [inline types/pdfmake/index.d.ts]: It looks like you did not add this type to the new interfaces.d.ts.

File: types/pdfmake/interfaces.d.ts [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+    table?: Table;
+    image?: string;
+    qr?: string;
+    canvas?: CanvasElement;
+    svg?: string;
+    columns?: Column[];
+    id?: string;
+    headlineLevel?: number;
+    style?: string | string[] | Style;
+    pageBreak?: 'before' | 'after';
```
Comment (nmummau) [inline types/pdfmake/interfaces.d.ts]: why not create a new type for this instead? That's how it was previously defined.

File: types/pdfmake/interfaces.d.ts [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+}
+
+export interface TFontFamilyTypes {
+    normal?: string;
+    bold?: string;
+    italics?: string;
+    bolditalics?: string;
+}
+
+export interface TDocumentInformation {
```
Comment (nmummau) [inline types/pdfmake/interfaces.d.ts]: Please add back type annotation comments.

File: types/pdfmake/interfaces.d.ts [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+    };
+}
+
+export interface StyleDictionnary {
+    [name: string]: Style;
+}
+
+export type PDFVersion = '1.3' | '1.4' | '1.5' | '1.6' | '1.7' | '1.7ext3';
+
+export interface Watermark {
```
Comment (nmummau) [inline types/pdfmake/interfaces.d.ts]: Please add back type annotation comments.

File: types/pdfmake/interfaces.d.ts [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+export interface ContentOrderedList extends ContentBase {
+    ol: OrderedListElement[];
+    type?: OrderedListType;
+    markerColor?: string;
+    separator?: string | [string, string];
+    reversed?: boolean;
+    start?: number;
+}
+
+export interface ContentUnorderedList extends ContentBase {
```
Comment (nmummau) [inline types/pdfmake/interfaces.d.ts]: Please add back type annotation comments.

File: types/pdfmake/interfaces.d.ts [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+
+export interface ContentColumns extends ContentBase {
+    columns: Column[];
+}
+
+export interface ContentStack extends ContentBase {
+    stack: Content[];
+}
+
+export interface ContentOrderedList extends ContentBase {
```
Comment (nmummau) [inline types/pdfmake/interfaces.d.ts]: Please add back type annotation comments.

File: types/pdfmake/index.d.ts (hunk 2 above)
Comment (jeralm) [reply types/pdfmake/index.d.ts]: I did here: https://github.com/jeralm/DefinitelyTyped/blob/a542e86b76952f8459e0a23b650358cbd9d6d43c/types/pdfmake/interfaces.d.ts#L61

(Is this the best way to link code? I'm not too experienced with github)

File: types/pdfmake/index.d.ts (hunk 3 above)
Comment (jeralm) [reply types/pdfmake/index.d.ts]: I did here: https://github.com/jeralm/DefinitelyTyped/blob/a542e86b76952f8459e0a23b650358cbd9d6d43c/types/pdfmake/interfaces.d.ts#L155

File: types/pdfmake/interfaces.d.ts (hunk 4 above)
Comment (jeralm) [reply types/pdfmake/interfaces.d.ts]: You're right, I'll update it

File: types/pdfmake/interfaces.d.ts (hunk 5 above)
Comment (jeralm) [reply types/pdfmake/interfaces.d.ts]: I didn't pay attention while merging, will do.

File: types/pdfmake/interfaces.d.ts (hunk 6 above)
Comment (jeralm) [reply types/pdfmake/interfaces.d.ts]: I didn't pay attention while merging, will do.

File: types/pdfmake/interfaces.d.ts (hunk 7 above)
Comment (jeralm) [reply types/pdfmake/interfaces.d.ts]: I didn't pay attention while merging, will do.

File: types/pdfmake/interfaces.d.ts (hunk 8 above)
Comment (jeralm) [reply types/pdfmake/interfaces.d.ts]: I didn't pay attention while merging, will do.

File: types/pdfmake/interfaces.d.ts (hunk 1 above)
Comment (jeralm) [reply types/pdfmake/interfaces.d.ts]: I didn't pay attention while merging, will do.

**Comment by jeralm** [comment]
@nmummau I added back everything you recommended.

File: types/pdfmake/interfaces.d.ts (hunk 1 above)
Comment (nmummau) [reply types/pdfmake/interfaces.d.ts]: This looks good.

File: types/pdfmake/index.d.ts (hunk 3 above)
Comment (nmummau) [reply types/pdfmake/index.d.ts]: This looks good, thank you.

File: types/pdfmake/interfaces.d.ts (hunk 4 above)
Comment (nmummau) [reply types/pdfmake/interfaces.d.ts]: This looks good.

File: types/pdfmake/interfaces.d.ts (hunk 5 above)
Comment (nmummau) [reply types/pdfmake/interfaces.d.ts]: This looks good.

File: types/pdfmake/interfaces.d.ts (hunk 6 above)
Comment (nmummau) [reply types/pdfmake/interfaces.d.ts]: This looks good.

File: types/pdfmake/interfaces.d.ts (hunk 7 above)
Comment (nmummau) [reply types/pdfmake/interfaces.d.ts]: This looks good.

File: types/pdfmake/interfaces.d.ts (hunk 8 above)
Comment (nmummau) [reply types/pdfmake/interfaces.d.ts]: This looks good.

**Review by nmummau — CHANGES_REQUESTED** [review]
"Dictionnary" is spelled incorrectly throughout this PR.

**Comment by jeralm** [comment]
> "Dictionnary" is spelled incorrectly throughout this PR.

Oof, pardon my literal french. Fixed.

**Comment by nmummau** [comment]
@jeralm how do I test this locally? I would like to add your types to my project to make sure all is well. Let me know. This is the first time I'm reviewing such a large types definition change.

**Comment by jeralm** [comment]
@nmummau afaik you can add other type roots in your tsconfig file, and dump the pdfmake folder there. See [the docs](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types). You might have to remove / disable the pdfmake typedefs from your node_modules.

Do note that this is absolutely a breaking change in regards to the previous type definitions (as they were incomplete / incorrect). Your existing usage of pdfmake will probably throw some compilation errors, and debugging those is a bit tedious for large documents. Best of luck!

**Comment by nmummau** [comment]
> @nmummau afaik you can add other type roots in your tsconfig file, and dump the pdfmake folder there. See [the docs](https://www.typescriptlang.org/docs/handbook/tsconfig-json.html#types-typeroots-and-types). You might have to remove / disable the pdfmake typedefs from your node_modules.
> 
> Do note that this is absolutely a breaking change in regards to the previous type definitions (as they were incomplete / incorrect). Your existing usage of pdfmake will probably throw some compilation errors, and debugging those is a bit tedious for large documents. Best of luck!

And I will have to `npm i @types/pdfkit -D` to make that work, right?

**Comment by nmummau** [comment]
Additionally, @jeralm how do I change my imports to support this change? I expect other users to ask this same question since this is a breaking change.

Currently I am doing the following, which no longer works:
```
import * as pdfMake from 'pdfmake/build/pdfmake';
import * as pdfFonts from 'pdfmake/build/vfs_fonts';

const pdf = pdfMake;
pdf.vfs = pdfFonts.pdfMake.vfs;

**Comment by jeralm** [comment]
You need pdfkit only for your local tests.

As for the imports, here's examples for all scenarios:

```typescript
// Importing pdfMake with a script tag:
// <script src='build/pdfmake.min.js'></script>
// <script src='build/vfs_fonts.js'></script>

const fonts = {
    Roboto: {
        normal: 'Roboto-Regular.ttf',
        bold: 'Roboto-Medium.ttf',
        italics: 'Roboto-Italic.ttf',
        bolditalics: 'Roboto-MediumItalic.ttf',
    },
};

pdfMake.fonts = fonts;

const dd = {
    content: 'Hello world!',
};

pdfMake.createPdf(dd);

pdfMake.createPdf(dd, {}, fonts);
```

```typescript
// Importing pdfMake with a synthesized default import
// Note: this requires the --allowSyntheticDefaultImports or --esModuleInterop compiler options enabled

pdfMake.vfs = pdfFonts.pdfMake.vfs;

const content: ContentText = {
    text: 'Hello world!'
};

const dd: TDocumentDefinitions = {
    content
};

const document: TCreatedPdf = pdfMake.createPdf(dd);
document.open();
```

```typescript
// Importing pdfMake as a CommonJS module

import pdfFonts = require('pdfmake/build/vfs_fonts');
import pdfMake = require('pdfmake/build/pdfmake');
import { ContentText, TDocumentDefinitions } from 'pdfmake/interfaces';

pdfMake.vfs = pdfFonts.pdfMake.vfs;

const content: ContentText = {
    text: 'Hello world!'
};

const dd: TDocumentDefinitions = {
    content
};

const document: pdfMake.TCreatedPdf = pdfMake.createPdf(dd);
document.open();
```

```typescript
// Importing pdfMake in node (server-side)

import PdfPrinter = require('pdfmake');
import { TDocumentDefinitions, TFontDictionary } from 'pdfmake/interfaces';

const fonts: TFontDictionary = {
    Roboto: {
        normal: 'fonts/Roboto-Regular.ttf',
        bold: 'fonts/Roboto-Medium.ttf',
        italics: 'fonts/Roboto-Italic.ttf',
        bolditalics: 'fonts/Roboto-MediumItalic.ttf',
    },
};

const printer = new PdfPrinter(fonts);

const dd: TDocumentDefinitions = {
    content: 'Hello world!',
};

printer.createPdfKitDocument(dd);
```

The `pdfmake/interface` type module was inspired from [uuid](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/5a59a59b867088e492e59e9f52ec681a9ff5b7a2/types/uuid/interfaces.d.ts).

**Comment by nmummau** [comment]
Why do we have to explicitly import each interface that we want from interfaces.d.ts? Wouldn't it be much better if all interfaces were exported from pdfmake.d.ts? That way we could do something like `pdfmake.ContentText = {}` and `pdfmake.TDocumentDefinitions = {}`?

I do have `"allowSyntheticDefaultImports": true` set in my tsconfig.json.
I am getting an error when setting `pdfMake.vfs = pdfFonts.pdfMake.vfs;`.
```
let vfs: TFontDictionary
Cannot assign to 'vfs' because it is a read-only property.ts(2540)
```

image

**Comment by jeralm** [comment]
I read a lot on modules to try to find the correct way of defining types for pdfMake. A lot of it is still very confusing to me, but here's what I understand:

There is nothing stopping you from doing `import * as pdfTypes from 'pdfmake/interfaces';`, if that's more your style. The problem lies with pdfMake not respecting the ES6 module spec.

Browser:

According to the ES6 spec, a namespace import (`import * as foo from 'foo';`) gives a plain (uncallable) read-only object. TypeScript is not strict on that point to allow inter-operation with other module systems such as UMD (used by pdfMake). To get the strict behavior, [you must use the --esModuleInterop compiler option, which is recommended by TypeScript for all projects](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#support-for-import-d-from-cjs-from-commonjs-modules-with---esmoduleinterop).*

You are getting the error because of the namespace import. To import a mutable object, you HAVE to use the `import pdfMake = require('pdfMake');` or `import pdfMake from 'pdfmake';` syntax, to mark them as non-ES6 modules.

Server:

The server side module exports a single callable object (the `PdfPrinter` class). This forces the use of the `export = PdfPrinter` syntax in the server side typedefs, which forbids any other exports. I think there might be a way to do some [declaration merging](https://www.typescriptlang.org/docs/handbook/declaration-merging.html) magic to merge a namespace with the class, but I don't think doing `let doc: PdfPrinter.TDocumentDefinitions;` is very ergonomic, and you still wouldn't be able to do do a namespace import.

Conclusion:

I am hesitant to add your recommendation because I think it's best to have the experience be uniform across the server and browser. I also think it's best not to export symbols from multiple files at the same time. The way it is now, `pdfmake/interfaces` is a compliant ES6 module, and the actual pdfmake api is whatever it is. I do agree however that regression is undesirable. I'll try it in a branch and see if it makes sense.

Notes:
*It looks like the --allowSyntheticDefaultImports options also has this strict behavior. I do not know if this is documented anywhere.

**Comment by jeralm** [comment]
@nmummau [It's not pretty and the backend somehow requires TypeScript version 3.7](https://github.com/jeralm/DefinitelyTyped/blob/d795e29138469aa7dceb488aa3a9d4c47bec6834/types/pdfmake/index.d.ts#L10)

I could do only the change you suggested in the browser, and leave the server as it is, but I'd personnally rather not to export the same symbols from multiple files. I'll let you choose what's better, as I have never used pdfMake in the front-end.

**Comment by jeralm** [comment]
@nmummau Could you give some feedback? I'd really like to get this PR over with

**Comment by nmummau** [comment]
I would appreciate if some of the other contributors could weigh in
@m1llen1um @radziksh @evolkmann @andipaetzold

**Comment by jeralm** [comment]
@nmummau They have not responded for a whole 8 days (even more considering my previous PR). If you approve the changes, a TypeScript or DefinitelyTyped team member will review the PR before merging.

**Comment by nmummau** [comment]
Approved, but this is a breaking change and should be a major version bump in DefinitelyTyped for types/pdfmake

**Review by nmummau — APPROVED** [review]

**Comment by jeralm** [comment]
Is there a reason why this PR hasn't been merged yet? Or is this the usual wait time for large PRs? Is it because it still has the "Revision Needed" tag?

**Comment by uniqueiniquity** [comment]
Hi @jeralm, I apologize for the delay.
I'll be able to review and merge this soon.

**Comment by jeralm** [comment]
No problem, take your time, I was just worried something fell through the cracks

**Comment by uniqueiniquity** [comment]
Congratulations on your first DefinitelyTyped contribution!
Thank you for being a part of the community!

**Comment by athencxss** [comment]
Since this is a breaking change might've been nice to change the versioning to a major so that everyone doesn't pull in this and break their code.

**Comment by jeralm** [comment]
@athencxss As far as I know, that is why the patch version of @types packages is managed by DefinitelyTyped. You can still use [the previous version (0.1.11)](https://www.npmjs.com/package/@types/pdfmake/v/0.1.11) of the type definitions.

This was a special case where the previous type definitions were *incorrect* and had to be fixed with a breaking change, but the library itself never put out any breaking changes. As such, I don't think I could upgrade the minor version in this typedef, as it wouldn't correspond to the library's minor version.

The contribution guide did not have guidelines for this kind of contributions, so I figured this was the way to go, and hoped a DefinitelyTyped maintainer would correct me if I was wrong.

@uniqueiniquity Can you confirm my understanding?

**Comment by uniqueiniquity** [comment]
Yes, your understanding is correct.
In a sense, most changes to types packages are breaking, because their job is to break your build so it doesn't break at runtime.
Therefore, the package version should not be versioned separately from the underlying package's version, and thus, you did the right thing. :)

**Comment by princemaple** [comment]
Hi,

I just tried adding pdfmake types to my project which is already using pdfmake, while using it as `any`. I found a couple places where the types don't match the real use case.

```ts
type TableLayout = PredefinedTableLayout | CustomTableLayout;
```

should probably be `PredefinedTableLayout | {[key: string]: CustomTableLayout}`

Reference from docs:
https://pdfmake.github.io/docs/document-definition-object/tables/
```js
pdfMake.tableLayouts = {
  exampleLayout: {
    hLineWidth: function (i, node) {
      if (i === 0 || i === node.table.body.length) {
        return 0;
      }
      return (i === node.table.headerRows) ? 2 : 1;
    },
    vLineWidth: function (i) {
      return 0;
    },
    hLineColor: function (i) {
      return i === 1 ? 'black' : '#aaa';
    },
    paddingLeft: function (i) {
      return i === 0 ? 0 : 8;
    },
    paddingRight: function (i, node) {
      return (i === node.table.widths.length - 1) ? 0 : 8;
    }
  }
};
```

```ts
export function createPdf(
    documentDefinitions: TDocumentDefinitions,
    tableLayouts?: TableLayout,
    fonts?: TFontDictionary,
    vfs?: TFontDictionary,  // <- this
): TCreatedPdf;
```

might be wrong as well, vfs type is different from fonts type, hence should not be TFontDictionary.

vfs looks like this
```js
{
  "Roboto-Italic.ttf": "AAEAAAASAQAABAAgR0RFRrRCsIIAAjGsAAA....base64 of font binary",
  "Roboto-Medium.ttf": "AAEAAAASAQAABAAgR0RFRrRCsIIAAiL8AAA....",
  "Roboto-MediumItalic.ttf": "AAEAAAASAQAABAAgR0RFRrRCsIIAAjW8AAAC....",
  "Roboto-Regular.ttf": "AAEAAAASAQAABAAgR0RFRrRCsIIAAijcAA...."
}
```

should probably just be any flat object.

**Comment by jeralm** [comment]
`TFontDictionary` is really a `{ [key: string]: string }`, so it does correspond to the type of `pdfMake.vfs`, although maybe a little too narrow in its description. I wish there was a way to type check the shape of a base64 string, but there isn't.

As for the tableLayouts, you are correct. `pdfMale

**Comment by jeralm** [comment]
You are correct. `pdfMake.tableLayouts`, `BufferOptions.tableLayouts` and the `tableLayouts` parameter of `pdfMake.createPdf()` should all be `{ [name: string]: CustomTableLayout }`. `type TableLayout`, which is used in `ContentTable`, should be `PredefinedTableLayout | string | CustomTableLayout` to allow usage of custom defined table layouts.

As for vfs, it seems you are also correct.

If you feel like fixing it yourself, open a PR, I will gladly review it. Don't forget tests. Otherwise I might have some time to do the fix myself tomorrow night. Either way, you should open an issue to track the issue.

**Comment by princemaple** [comment]
> Otherwise I might have some time to do the fix myself tomorrow night. Either way, you should open an issue to track the issue.

Opened #43758. I probably won't be able to get to this before you do. Thank you for your great effort.

**Pull request merged**